### PR TITLE
*: Fix possible undefined behavior for transmuting vec

### LINKED ISSUE #12070: UB in TsSet conversions
https://github.com/tikv/tikv/blob/8b86c063c21f111a2fb0d4fcc4a328c3df7bd6f3/components/txn_types/src/timestamp.rs#L149
https://github.com/tikv/tikv/blob/8b86c063c21f111a2fb0d4fcc4a328c3df7bd6f3/components/txn_types/src/timestamp.rs#L156

This unsafe cast is UB because Vec\<u64\> and Vec\<TimeStamp\> don't have the same layout in general. For example a Rust implementation using profile-guided optimization might put the vec's fields in a different order for the two types, due to different access patterns across all the places those 2 different instantiations are used (e.g. one might store the len first while the other might store the capacity first, because len() is called more often on one than the other).

### PR DESCRIPTION
**Pull request by Connor1996**
### What is changed and how it works?

Issue Number: Close #12070

What's Changed:

Recreate a vec from raw parts instead of transmuting memory directly.

### Related changes

- Need to cherry-pick to the release branch

### Check List 

Tests 

- No code

### Release note 

```release-note
Fix possible undefined behavior for transmuting vec
```

### CODE REVIEW

**Review by sticnarf — APPROVED** [review]
LGTM

File: components/concurrency_manager/src/lib.rs [hunk 1]
```diff
@@ -78,11 +80,8 @@ impl ConcurrencyManager {
         for (index, key) in keys_with_index {
             result[index] = MaybeUninit::new(self.lock_table.lock_key(key).await);
         }
-        #[allow(clippy::unsound_collection_transmute)]
-        #[allow(clippy::transmute_undefined_repr)]
-        unsafe {
-            mem::transmute(result)
-        }
+        let (ptr, len, cap) = result.into_raw_parts();
```
Comment (BusyJay) [inline components/concurrency_manager/src/lib.rs:83]: How about making it a util function?

File: components/tikv_util/src/memory.rs [hunk 2]
```diff
@@ -8,6 +8,13 @@ use kvproto::{
 };
 use std::mem;
 
+// Transmute vec with types of same memory layout.
+pub unsafe fn vec_transmute<F, T>(from: Vec<F>) -> Vec<T> {
+    debug_assert!(mem::size_of::<F>() == mem::size_of::<T>());
```
Comment (sticnarf) [inline components/tikv_util/src/memory.rs:13]: Better to also assert about the alignment.

And `static_assertions` may help check the error in compile time.

Comment (BusyJay) [inline components/tikv_util/src/memory.rs:13]: May also check for align.

File: components/tikv_util/src/memory.rs [hunk 3]
```diff
@@ -8,6 +8,13 @@ use kvproto::{
 };
 use std::mem;
 
+// Transmute vec with types of same memory layout.
+pub unsafe fn vec_transmute<F, T>(from: Vec<F>) -> Vec<T> {
```
Comment (BusyJay) [inline components/tikv_util/src/memory.rs:12]: How about marking it `inline`?

**Review by BusyJay — APPROVED** [review]

File: components/tikv_util/src/memory.rs (hunk 2 above)
Comment (Connor1996) [reply components/tikv_util/src/memory.rs:14]: can't, rust complains about it. Same goes for `static_assertions:: assert_eq_size`.
```
error[E0401]: can't use generic parameters from outer function
  --> components/tikv_util/src/memory.rs:14:56
   |
13 | pub unsafe fn vec_transmute<F, T>(from: Vec<F>) -> Vec<T> {
   |                             - type parameter from outer function
14 |     static_assertions::const_assert_eq!(mem::size_of::<F>(), mem::size_of::<T>());
   |                                                        ^ use of generic parameter from outer function
```
Seems no perfect solution for this, see https://stackoverflow.com/questions/30330519/compile-time-generic-type-size-check

Comment (sticnarf) [reply components/tikv_util/src/memory.rs:14]: OK, I see.

**Pull request merged**